File: Medium/983. Minimum cost for tickets/soln.java
public class Node
{
    int i;
    List<Node> adj;
    List<Integer> cost;
    int dist;
    Node(int x)
    {
        i = x;
        adj = new ArrayList<>();
        cost = new ArrayList<>();
        dist = Integer.MAX_VALUE;
    }
}
class Solution
{
    public int mincostTickets(int[] days, int[] costs)
    {
        Node[] graph = new Node[days.length+1];
        for(int i=0;i<graph.length;i++)
        {
            graph[i] = new Node(i<days.length?days[i]:-1);
        }
        graph[0].dist=0;
        for(int i=0;i<graph.length-1;i++)
        {
            graph[i].adj.add(graph[i+1]);
            graph[i].cost.add(Math.min(costs[0],Math.min(costs[1],costs[2])));
            for(int j=i+2; j<graph.length; j++)
            {
                if(days[j-1]-days[i]<7)
                {
                    graph[i].adj.add(graph[j]);
                    graph[i].cost.add(Math.min(costs[1],costs[2]));
                }
                else if(days[j-1]-days[i]<30)
                {
                    graph[i].adj.add(graph[j]);
                    graph[i].cost.add(costs[2]);
                }
                else
                {
                    break;
                }
            }
        }
        for(int i = 0; i<graph.length; i++)
        {
            for(int j=0; j<graph[i].adj.size(); j++)
            {
                graph[i].adj.get(j).dist = Math.min(graph[i].adj.get(j).dist ,
                                               graph[i].dist + graph[i].cost.get(j));
            }
        }
        return graph[days.length].dist;
    }
}

File: Hard/42. Rainwater Trapping/soln.cpp
class Solution
{
public:
    int trap(vector<int>& a)
    {
        int i=0,j=0,k=0,m=0,s1=0,s=0,s2=0,n;
        n=a.size();
        if (n<=2)
            return 0;
        if (a[0]<=a[1])
        {    
            while (a[i]<=a[i+1])
            {
                //n=n-1;
                i++;
                if (n<=2)
                    return 0;
            }
        }
        if (a[n-2]>=a[n-1])
        {    
            while (n>1 && a[n-2]>=a[n-1])
            {
                n=n-1;
                if (n<=2)
                    return 0;
            }
        }
        if (n<=2)
            return 0;
        k=n-1;
        while (i<n || k>=0)
        {
            j=i+1;
            m=k-1;
            s1=0;
            s2=0;
            while (j<n && i<n)
            {
                if (a[j]<a[i])
                    s1=s1+a[i]-a[j];
                else
                {
                    s=s+s1;
                    i=j;
                    break;
                }
                j=j+1;
            }
            i=j;
            while (m>=0 && k>=0)
            {
                if (a[m]<=a[k])
                    s2=s2+a[k]-a[m];
                else
                {   
                    s=s+s2;
                    k=m;
                    break;
                }
                m=m-1;
            }
            k=m;
        }
        return s;
    }
};